### PR TITLE
#356 Part I -- Move Image Upload API

### PR DESCRIPTION
**Pull request by chriswininger**
I am preparing to implement a first pass of basic image per-processing which will make sure we do not store images that are too large. As an initial first step I have moved image processing out the web-portal and into the API-Server where we are supported by the more robust backend functionality that NextJs provieds.

This inadvertanly addresses issue #436. In the re-write I have not ported the bug :-)

### CODE REVIEW

File: api-server/src/constants.ts [hunk 1]
```diff
@@ -7,8 +7,10 @@
  */
 import fs from 'fs';
 import path from 'path';
-import { NullableOrUndefinable } from './utils/NullableOrUndefinable';
+import { Nullable, NullableOrUndefinable } from './utils/NullableOrUndefinable';
 import { isNullOrUndefined } from './utils';
+import * as process from 'process';
```
Comment (MatthewGidcomb) [inline api-server/src/constants.ts]: Why does this need to be imported?

File: api-server/src/uploads/uploads.controller.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  ): Promise<UploadsResponse> {
+    const imagePath = await this.uploadsService.saveImage(file);
+    return new UploadsResponse({ imagePath });
+  }
+}
+
+function getImageUploadValidators(): ParseFilePipe {
+  return new ParseFilePipeBuilder()
+    .addFileTypeValidator({
+      fileType: /image\/(jpe?g|png|webp|tif|tiff|bmp)$/i,
```
Comment (MatthewGidcomb) [inline api-server/src/uploads/uploads.controller.ts]: Does this support `image/*`? That's what the input on the form uses for `accept` (per [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#unique_file_type_specifiers) it should accept any image format).

(I wish Nestjs had API docs to supplement the higher-level ones.)

Comment (chriswininger) [reply api-server/src/uploads/uploads.controller.ts]: yeah this was quite frustrating to get right what I'm matching on are mime types such as image/jpeg, image/png. It's just a regular expression so you could match against image/*, that would be `/image\/.*$/i` I believe.

If that's what we're allowing in the form that would make sense though long term we should probably lock this down more. I doubt we want to support any image format

File: api-server/src/constants.ts (hunk 1 above)
Comment (chriswininger) [reply api-server/src/constants.ts]: do you mean the line below this? process is being used, though the import below this `import isNotNullOrUndefined from './utils/is-not-null-or-undefined';` is unused and I will remove it

Comment (chriswininger) [reply api-server/src/constants.ts]: 86b80a3dbd05d542c4145ac92684e08f184a8e35

Comment (chriswininger) [reply api-server/src/constants.ts]: 🤔 maybe though you did mean process looks like it an work without importing it, guess it's special, doesn't seem to matter either way

Comment (chriswininger) [reply api-server/src/constants.ts]: removed 253ceecb7727632ef08d1ab92774ce4343a7bba9

File: api-server/src/uploads/uploads.controller.ts (hunk 2 above)
Comment (chriswininger) [reply api-server/src/uploads/uploads.controller.ts]: loosened up the requirements 9bf391a22a4a612f87b00a114fc108ff9f614b97

**Pull request merged**